### ISSUE TITLE
fs-repo-migrations doesn't appear to work in alpine containers

### ISSUE THREAD

**Issue by evanstucker-hates-2fa** [ipfs/fs-repo-migrations#121]
I get a "not found" error when I run this in an Alpine Linux container:
```
$ docker run -it --rm alpine sh
/ # apk add curl
...
/ # curl -L https://dist.ipfs.io/fs-repo-migrations/v1.7.1/fs-repo-migrations_v1.7.1_linux-amd64.tar.gz | tar -xz
...
/ # cd fs-repo-migrations
/fs-repo-migrations # ./fs-repo-migrations
sh: ./fs-repo-migrations: not found
/fs-repo-migrations # 
```
But following the same steps, it works in Ubuntu:
```
$ docker run -it --rm ubuntu sh
# apt update
...
# apt install curl
...
# curl -L https://dist.ipfs.io/fs-repo-migrations/v1.7.1/fs-repo-migrations_v1.7.1_linux-amd64.tar.gz | tar -xz
# cd fs-repo-migrations
# ./fs-repo-migrations
ipfs migration:  stat /root/.ipfs: no such file or directory
```
This is presumably because Alpine uses musl instead of glibc.

Can we either provide a binary that works on Alpine, or improve the "not found" error message to something like "This requires glibc. It won't work on Alpine Linux."

**Comment by aschmahmann** [comment]
>This is presumably because Alpine uses musl instead of glibc.

Yes.

> Can we either provide a binary that works on Alpine, or improve the "not found" error message to something like "This requires glibc. It won't work on Alpine Linux."

You get a musl-related error when trying to upgrade using go-ipfs itself although I guess adding more checks in here could be done as well.

Out of curiosity/trying to understand user flow, why did you end up manually downloading the binary from dist?

**Comment by evanstucker-hates-2fa** [comment]
I am running the linuxserver.io IPFS image and I was just trying to grab a binary to do the migration inside an existing docker image. See more here: https://github.com/linuxserver/docker-ipfs/issues/4

I did try the "go get" method inside that image, but I got an error:
```
$ docker exec -it ipfs sh
root@cdbd939a571c:/# apk add git go
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/community/x86_64/APKINDEX.tar.gz
(1/15) Installing expat (2.2.9-r1)
(2/15) Installing git (2.26.2-r0)
(3/15) Installing libgcc (9.3.0-r2)
(4/15) Installing libstdc++ (9.3.0-r2)
(5/15) Installing binutils (2.34-r1)
(6/15) Installing gmp (6.2.0-r0)
(7/15) Installing isl (0.18-r0)
(8/15) Installing libgomp (9.3.0-r2)
(9/15) Installing libatomic (9.3.0-r2)
(10/15) Installing libgphobos (9.3.0-r2)
(11/15) Installing mpfr4 (4.0.2-r4)
(12/15) Installing mpc1 (1.1.0-r1)
(13/15) Installing gcc (9.3.0-r2)
(14/15) Installing musl-dev (1.1.24-r10)
(15/15) Installing go (1.13.15-r0)
Executing busybox-1.31.1-r19.trigger
OK: 531 MiB in 61 packages
root@cdbd939a571c:/# go get -u github.com/ipfs/fs-repo-migrations
# github.com/ipfs/fs-repo-migrations/ipfs-10-to-11/_vendor/github.com/lucas-clemente/quic-go/internal/qtls
root/go/src/github.com/ipfs/fs-repo-migrations/ipfs-10-to-11/_vendor/github.com/lucas-clemente/quic-go/internal/qtls/qtls.go:79:25: cannot convert cert (type *tls.Certificate) to type *qtls.Certificate
root@cdbd939a571c:/# 
```
This is really weird... I see that it works fine in a base alpine image:
```
$ docker run -it --rm alpine sh
/ # go get -u github.com/ipfs/fs-repo-migrations
sh: go: not found
/ # apk add go
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/13) Installing libgcc (10.2.1_pre1-r3)
(2/13) Installing libstdc++ (10.2.1_pre1-r3)
(3/13) Installing binutils (2.35.1-r1)
(4/13) Installing libgomp (10.2.1_pre1-r3)
(5/13) Installing libatomic (10.2.1_pre1-r3)
(6/13) Installing libgphobos (10.2.1_pre1-r3)
(7/13) Installing gmp (6.2.1-r0)
(8/13) Installing isl22 (0.22-r0)
(9/13) Installing mpfr4 (4.1.0-r0)
(10/13) Installing mpc1 (1.2.0-r0)
(11/13) Installing gcc (10.2.1_pre1-r3)
(12/13) Installing musl-dev (1.2.2-r0)
(13/13) Installing go (1.15.7-r0)
Executing busybox-1.32.1-r2.trigger
OK: 436 MiB in 27 packages
/ # go get -u github.com/ipfs/fs-repo-migrations
go: missing Git command. See https://golang.org/s/gogetcmd
package github.com/ipfs/fs-repo-migrations: exec: "git": executable file not found in $PATH
/ # apk add git
(1/7) Installing ca-certificates (20191127-r5)
(2/7) Installing brotli-libs (1.0.9-r3)
(3/7) Installing nghttp2-libs (1.42.0-r1)
(4/7) Installing libcurl (7.74.0-r0)
(5/7) Installing expat (2.2.10-r1)
(6/7) Installing pcre2 (10.36-r0)
(7/7) Installing git (2.30.1-r0)
Executing busybox-1.32.1-r2.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 449 MiB in 34 packages
/ # go get -u github.com/ipfs/fs-repo-migrations
/ # which fs-repo-migrations
/ # find . -name fs-repo-migrations
./root/go/bin/fs-repo-migrations
./root/go/src/github.com/ipfs/fs-repo-migrations
/ # ./root/go/bin/fs-repo-migrations
ipfs migration:  stat /root/.ipfs: no such file or directory
/ # 
```

**Comment by evanstucker-hates-2fa** [comment]
Ah ha! It's a problem with alpine:3.12!
```
$ docker run -it --rm alpine:3.12 sh
Unable to find image 'alpine:3.12' locally
3.12: Pulling from library/alpine
f84cab65f19f: Pull complete 
Digest: sha256:a295107679b0d92cb70145fc18fb53c76e79fceed7e1cf10ed763c7c102c5ebe
Status: Downloaded newer image for alpine:3.12
/ # apk add git go
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.12/community/x86_64/APKINDEX.tar.gz
(1/19) Installing ca-certificates (20191127-r4)
(2/19) Installing nghttp2-libs (1.41.0-r0)
(3/19) Installing libcurl (7.69.1-r3)
(4/19) Installing expat (2.2.9-r1)
(5/19) Installing pcre2 (10.35-r0)
(6/19) Installing git (2.26.2-r0)
(7/19) Installing libgcc (9.3.0-r2)
(8/19) Installing libstdc++ (9.3.0-r2)
(9/19) Installing binutils (2.34-r1)
(10/19) Installing gmp (6.2.0-r0)
(11/19) Installing isl (0.18-r0)
(12/19) Installing libgomp (9.3.0-r2)
(13/19) Installing libatomic (9.3.0-r2)
(14/19) Installing libgphobos (9.3.0-r2)
(15/19) Installing mpfr4 (4.0.2-r4)
(16/19) Installing mpc1 (1.1.0-r1)
(17/19) Installing gcc (9.3.0-r2)
(18/19) Installing musl-dev (1.1.24-r10)
(19/19) Installing go (1.13.15-r0)
Executing busybox-1.31.1-r19.trigger
Executing ca-certificates-20191127-r4.trigger
OK: 458 MiB in 33 packages
/ # go get -u github.com/ipfs/fs-repo-migrations
# github.com/ipfs/fs-repo-migrations/ipfs-10-to-11/_vendor/github.com/lucas-clemente/quic-go/internal/qtls
root/go/src/github.com/ipfs/fs-repo-migrations/ipfs-10-to-11/_vendor/github.com/lucas-clemente/quic-go/internal/qtls/qtls.go:79:25: cannot convert cert (type *tls.Certificate) to type *qtls.Certificate
/ # 
```
It's works correctly in alpine:3.13:
```
$ docker run -it --rm alpine:3.13 sh
Unable to find image 'alpine:3.13' locally
3.13: Pulling from library/alpine
ba3557a56b15: Already exists 
Digest: sha256:a75afd8b57e7f34e4dad8d65e2c7ba2e1975c795ce1ee22fa34f8cf46f96a3be
Status: Downloaded newer image for alpine:3.13
/ # apk add git go
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/20) Installing ca-certificates (20191127-r5)
(2/20) Installing brotli-libs (1.0.9-r3)
(3/20) Installing nghttp2-libs (1.42.0-r1)
(4/20) Installing libcurl (7.74.0-r0)
(5/20) Installing expat (2.2.10-r1)
(6/20) Installing pcre2 (10.36-r0)
(7/20) Installing git (2.30.1-r0)
(8/20) Installing libgcc (10.2.1_pre1-r3)
(9/20) Installing libstdc++ (10.2.1_pre1-r3)
(10/20) Installing binutils (2.35.1-r1)
(11/20) Installing libgomp (10.2.1_pre1-r3)
(12/20) Installing libatomic (10.2.1_pre1-r3)
(13/20) Installing libgphobos (10.2.1_pre1-r3)
(14/20) Installing gmp (6.2.1-r0)
(15/20) Installing isl22 (0.22-r0)
(16/20) Installing mpfr4 (4.1.0-r0)
(17/20) Installing mpc1 (1.2.0-r0)
(18/20) Installing gcc (10.2.1_pre1-r3)
(19/20) Installing musl-dev (1.2.2-r0)
(20/20) Installing go (1.15.7-r0)
Executing busybox-1.32.1-r3.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 449 MiB in 34 packages
/ # go get -u github.com/ipfs/fs-repo-migrations
/ # 
```

**Comment by evanstucker-hates-2fa** [comment]
Looks like my original issue has been solved by the maintainer upgrading to alpine:3.13: https://github.com/linuxserver/docker-ipfs/issues/4

You can close this issue. Just know that "go get" doesn't work for alpine:3.12... I don't know if this is a well known issue or not.

**Comment by gammazero** [comment]
Resolved as described in the previous comment.  Closing.
